pkg/middleware/middleware.go

```diff
@@ -26,21 +26,24 @@ func HandleNoCacheHeader(ctx *models.ReqContext) {
 }
 
 func AddDefaultResponseHeaders(cfg *setting.Cfg) web.Handler {
+	t := web.NewTree()
+	t.Add("/api/datasources/uid/:uid/resources/*", nil)
+	t.Add("/api/datasources/:id/resources/*", nil)
 	return func(c *web.Context) {
-		c.Resp.Before(func(w web.ResponseWriter) {
-			// if response has already been written, skip.
+		c.Resp.Before(func(w web.ResponseWriter) { // if response has already been written, skip.
 			if w.Written() {
 				return
 			}
 
-			if !strings.HasPrefix(c.Req.URL.Path, "/api/datasources/proxy/") {
+			_, _, resourceURLMatch := t.Match(c.Req.URL.Path)
+			resourceCachable := allowCacheControl(c.Resp) && resourceURLMatch
```

Comment:
Maybe check for match first make more sense?

```suggestion
			resourceCachable := resourceURLMatch && allowCacheControl(c.Resp)
```